source/credentials.c

```diff
@@ -213,34 +230,82 @@ struct aws_credentials_provider *aws_credentials_provider_new_chain_default(
         providers[index++] = profile_provider;
     }
 
-    struct aws_credentials_provider_imds_options imds_options;
-    AWS_ZERO_STRUCT(imds_options);
-    imds_options.bootstrap = options->bootstrap;
-    imds_provider = aws_credentials_provider_new_imds(allocator, &imds_options);
-    if (imds_provider == NULL) {
-        goto on_error;
-    }
+    if (ecs_relative_uri && ecs_relative_uri->len) {
+        struct aws_credentials_provider_ecs_options ecs_options;
+        AWS_ZERO_STRUCT(ecs_options);
+        ecs_options.bootstrap = options->bootstrap;
+        ecs_options.host = aws_byte_cursor_from_string(s_ecs_host);
+        ecs_options.path_and_query = aws_byte_cursor_from_string(ecs_relative_uri);
+        ecs_options.use_tls = false;
+        ecs_provider = aws_credentials_provider_new_ecs(allocator, &ecs_options);
+        if (ecs_provider == NULL) {
+            success = false;
+            goto on_ret;
+        }
+        providers[index] = ecs_provider;
+
+    } else if (ecs_full_uri && ecs_full_uri->len) {
+        struct aws_uri uri;
+        struct aws_byte_cursor uri_cstr = aws_byte_cursor_from_string(ecs_full_uri);
+        if (AWS_OP_ERR == aws_uri_init_parse(&uri, allocator, &uri_cstr)) {
+            success = false;
+            goto on_ret;
+        }
 
-    providers[index] = imds_provider;
+        struct aws_string *ecs_token = NULL;
+        aws_get_environment_value(allocator, s_ecs_creds_env_token, &ecs_token);
+        struct aws_credentials_provider_ecs_options ecs_options;
+        AWS_ZERO_STRUCT(ecs_options);
+        ecs_options.bootstrap = options->bootstrap;
+        ecs_options.host = uri.host_name;
+        ecs_options.path_and_query = uri.path_and_query;
+        ecs_options.use_tls = aws_byte_cursor_eq_c_str_ignore_case(&(uri.scheme), "HTTPS");
+        if (ecs_token->len) {
+            ecs_options.auth_token = aws_byte_cursor_from_string(ecs_token);
+        }
+        ecs_provider = aws_credentials_provider_new_ecs(allocator, &ecs_options);
+        aws_string_destroy(ecs_token);
+        if (ecs_provider == NULL) {
+            success = false;
+            goto on_ret;
+        }
+        providers[index] = ecs_provider;
+
+    } else if (ec2_imds_disable == NULL || aws_string_eq_c_str_ignore_case(ec2_imds_disable, "false")) {
+        struct aws_credentials_provider_imds_options imds_options;
+        AWS_ZERO_STRUCT(imds_options);
+        imds_options.bootstrap = options->bootstrap;
+        imds_provider = aws_credentials_provider_new_imds(allocator, &imds_options);
+        if (imds_provider == NULL) {
+            success = false;
+            goto on_ret;
+        }
+        providers[index] = imds_provider;
+    }
 
     AWS_FATAL_ASSERT(index < AWS_ARRAY_SIZE(providers));
 
     struct aws_credentials_provider_chain_options chain_options;
     AWS_ZERO_STRUCT(chain_options);
-    chain_options.provider_count = index;
+    chain_options.provider_count = index + 1;
```

Comment:
I think we should have just been doing `providers[index++] = x;` everywhere
which means you could revert this change,
but we should modify the assertion above to be `AWS_FATAL_ASSERT(index <= AWS_ARRAY_SIZE(providers));` instead of just `<`

Reply:
let's change it